### DISCUSSION TITLE
Volume mounts and host-path binds on vantage6

### DISCUSSION THREAD

**Discussion by hcadavid** [vantage6/v6-on-kubernetes-PoC#4] [Q&A] [answered]
To give context to the following questions, I need to describe my current understanding of how the I/O on the containerised algorithms work:

- When a new job is created, The node [creates a Docker volume for temporal data](https://github.com/vantage6/vantage6/blob/a054513a8fc8e4ca33f395faba6e503c70aa7c27/vantage6-node/vantage6/node/__init__.py#L336), which is expected to be reused by jobs with the same id within the same node (so they can reuse partial results of their 'parent' jobs).

- When the job execution [is requested](https://github.com/vantage6/vantage6/blob/a759f540c731a19d351c82b414b729b303ad7aa3/vantage6-node/vantage6/node/docker/task_manager.py#L189), the DockerTaskManager creates, on the host, a folder hierarchy for the input data, the JWS token, and the output. Then it creates the container with the algorithm, mounts the temporal data volume, and binds the folders (input/token/output) on it.

@bartvanb, @frankcorneliusmartin  Is this accurate? 

Given the above: 

- Is there a reason (e.g., privacy-related) for using Docker volumes for this temporal data, rather than using the same kind of host-paths bound to the container (as the ones used for input/output)? This is important as the concept of volume in Kubernetes differs from the one of Docker. The former is based on the 'persistent volume' abstraction (referring to any storage alternative, e.g. RAM, a block device, an NFS, etc). The latter refers to a proprietary, isolated filesystem managed by the Docker daemon. As far as I am concerned, this shared 'temporal' volume on Kubernetes can only be implemented through persistent volumes (something that may cause problems as discussed [here](https://github.com/bdh-generalization/v6-on-kubernetes-PoC/blob/b2e86911e7cd0f12fe3d85f98875fd479f378e85/node_poc/container_manager.py#L152)) or through host-paths bound to the POD/algorithms (the approach currently used on the PoC).

- On the DockerManager there is a [FIXME comment](https://github.com/vantage6/vantage6/blob/a759f540c731a19d351c82b414b729b303ad7aa3/vantage6-node/vantage6/node/docker/task_manager.py#L348) that points out a potential data leak with the way these I/O paths are bound to the container. However, I still can't grasp the details. Could you elaborate a bit on this, so we can keep it in mind for this PoC?

```
    def _make_task_folders(self) -> None:
        """ Generate task folders """
        # FIXME: We should have a separate mount/volume for this. At the
        #   moment this is a potential leak as containers might access input,
        #   output and token from other containers.
        #
        #   This was not possible yet as mounting volumes from containers
        #   is terrible when working from windows (as you have to convert
        #   from windows to unix several times...).
```

- Is there a description of the expected lifecycle of the temporal and I/O data, e.g., when is it cleared? From the code I understand that the temporal data and the generated folder hierarchies are [cleared only when the node is deleted](https://github.com/vantage6/vantage6/blob/fdaf7c3dcb653f89fd74f3ebb4d2cce3c80516ef/vantage6/vantage6/cli/node/remove.py#L45), am I right?

Thanks for your input on this!

**Comment by frankcorneliusmartin** [comment, marked as answer]
> When a new job is created, The node [creates a Docker volume for temporal data](https://github.com/vantage6/vantage6/blob/a054513a8fc8e4ca33f395faba6e503c70aa7c27/vantage6-node/vantage6/node/__init__.py#L336), which is expected to be reused by jobs with the same id within the same node (so they can reuse partial results of their 'parent' jobs).

✔️ , however this is subjected to change https://github.com/vantage6/vantage6/issues/154. Also we are considering to make these volumes based on sessions rather than on their `run_id`: https://github.com/vantage6/vantage6/issues/943#issuecomment-1938940858

When the job execution [is requested](https://github.com/vantage6/vantage6/blob/a759f540c731a19d351c82b414b729b303ad7aa3/vantage6-node/vantage6/node/docker/task_manager.py#L189), the DockerTaskManager creates, on the host, a folder hierarchy for the input data, the JWS token, and the output. Then it creates the container with the algorithm, mounts the temporal data volume, and binds the folders (input/token/output) on it.

 ✔️, but be aware that we also considering changing the mounts / environment vars based on the container action: https://github.com/vantage6/vantage6/issues/943#issuecomment-1938940858

> Is there a reason (e.g., privacy-related) for using Docker volumes for this temporal data, rather than using the same kind of host-paths bound to the container (as the ones used for input/output)?

No, this seemed like the easiest solution at the time as we could pass the volume along the containers. So both the node and algorithm containers could write to the same folder. We could narrow this down. Ideally this would be removed completely by https://github.com/vantage6/vantage6/issues/154. 

> On the DockerManager there is a [FIXME comment](https://github.com/vantage6/vantage6/blob/a759f540c731a19d351c82b414b729b303ad7aa3/vantage6-node/vantage6/node/docker/task_manager.py#L348) that points out a potential data leak with the way these I/O paths are bound to the container. However, I still can't grasp the details. Could you elaborate a bit on this, so we can keep it in mind for this PoC?

The mount is way to broad. Any container that comes in the node can potentially read all previous input, token en output files of other tasks. This would basically also be fixed by allowing algorithms to only access their input/output and token file.

> Is there a description of the expected lifecycle of the temporal and I/O data, e.g., when is it cleared? From the code I understand that the temporal data and the generated folder hierarchies are [cleared only when the node is deleted](https://github.com/vantage6/vantage6/blob/fdaf7c3dcb653f89fd74f3ebb4d2cce3c80516ef/vantage6/vantage6/cli/node/remove.py#L45), am I right?

No this has been discussed several times. There is another way: `v6 node clean`. Options that we delete them after X time automatically. When using sessions as described in https://github.com/vantage6/vantage6/issues/943#issuecomment-1938940858 the user should also be able to delete them.
